### DISCUSSION TITLE
when to use async?

### DISCUSSION THREAD

**Discussion by matthiasdebernardini** [antonilol/rust-bitcoincore-zmq#11] [Q&A] [answered]
I think I already know the answer but I am curious what to think, I have my app listening to core zmq on an inf loop. 

This way I can look at each incoming rawtx and check it to see if I have to take action.

Should I be using the async? Or rather, whats a good heuristic for deciding when to use async or not?

**Comment by matthiasdebernardini** [comment]
follow up: if your app is already using async, mine is for sqlx, then should you use async unless you have a good reason not to?

**Reply by antonilol** [reply]
using synchronous I/O will block the thread (the runtime thread), if you got an async app it is best to use async

**Comment by antonilol** [comment, marked as answer]
in general, when a synchronous program has to wait on I/O tasks, use async. polling core zmq in an infinite loop is a bit like an async runtime would poll the async tasks, but the async runtime will also be able to poll other I/O tasks like reading a file or sending something over the network, possibly using multiple threads to spread load.

if the whole program is woken up by zmq messages from bitcoin core and not (or rarely) doing other I/O things, sync if probably fine

**Reply by matthiasdebernardini** [reply]
I am doing rawtx, which are so few, that I don't think it matters either way. With blocks its even fewer.

This is a address watcher (checks for deposits), I am also doing postgres calls and bitcoincore rpc. 

I will keep it as is and if I need more performance in the future I can change it.

**Comment by matthiasdebernardini** [comment]
anyways, I ended up adding it just because and this is my code example

like you said, if your app is already async, may as well use it, got it working on first try

```rust
async fn run(zmq: &str) -> anyhow::Result<()> {
    let endpoints = &[zmq];
    let mut stream = bitcoincore_zmq::subscribe_async(endpoints).expect("zmq must work");
    info!("Listening for transactions through zmq @ {zmq}");
    while let Some(msg) = stream.next().await {
        match msg {
            Ok(msg) => {
                if let Err(e) = process_message(&msg).await {
                    anyhow::bail!("Could not process message {msg}: {e}")
                }
            }
            Err(e) => anyhow::bail!("Could not receive message: {e}"),
        }
    }

    Ok(())
}
```
